Add redirect print to tqdm for capa main

Aayush-Goel-04 opened the pull request:
<!--
Thank you for contributing to capa! <3

Please read capa's CONTRIBUTING guide if you haven't done so already.
It contains helpful information about how to contribute to capa. Check https://github.com/mandiant/capa/blob/master/.github/CONTRIBUTING.md

Please describe the changes in this pull request (PR). Include your motivation and context to help us review.

Please mention the issue your PR addresses (if any):
closes #issue_number
-->
Closes #749 

### Checklist

<!-- CHANGELOG.md has a `master (unreleased)` section. Please add bug fixes, new features, breaking changes and anything else you think is worthwhile mentioning in the release notes to this file. -->
- [x] No CHANGELOG update needed
<!-- Tests prove that your fix/work as expected and ensure it doesn't break on the feature. -->
- [x] No new tests needed
<!-- Please help us keeping capa documentation up-to-date -->
- [x] No documentation update needed

williballenthin reviewed (approved):
very nice

williballenthin commented:
would you add an entry in the changelog so we can recognize your work?

williballenthin commented:
can you also add a screenshot here showing the improved output?

mr-tz commented on capa/helpers.py:
```diff
@@ -85,6 +89,40 @@ def get_format(sample: str) -> str:
 
 
+@contextlib.contextmanager
+def redirecting_print_to_tqdm(disable_progress):
```
can you also use this function from the linter script?

williballenthin replied:
no, remove the function in the linter (https://github.com/mandiant/capa/blob/7ab083f19a77f7d6ab658a0f614e6154476aca17/scripts/lint.py#L870) and use the copy in helpers.py instead.

Aayush-Goel-04 replied:
Will make some changes

Aayush-Goel-04 commented:
> can you also add a screenshot here showing the improved output?

Hey @williballenthin,
After making the changes, I ran `pytest tests/test_main.py` to test the changes in main.py. The tests were successful without any issues. Could you please clarify the specific details you would like to see in the screenshot?

williballenthin commented:
> Could you please clarify the specific details you would like to see in the screenshot?

please run capa against some sample file, perhaps PMA 01-01.exe and show the output thats printed to the console.

Aayush-Goel-04 commented:
@williballenthin 
![Screenshot 2023-05-16 173133](https://github.com/mandiant/capa/assets/81844215/d586b510-5ef4-401b-8692-cb690967f0b7)

mr-tz commented:
This should improve the output when other things are logged along the execution, e.g., when running in debug mode `-d`.

Aayush-Goel-04 commented:
When running with debug mode
```
PS C:\Users\HP\Documents\GitHub\capa> capa -d '.\tests\data\Practical Malware Analysis Lab 01-01.exe_'
DEBUG:capa:--------------------------------------------------------------------------------
DEBUG:capa: Using default embedded rules.
DEBUG:capa: To provide your own rules, use the form `capa.exe -r ./path/to/rules/  /path/to/mal.exe`.
DEBUG:capa: You can see the current default rule set here:
DEBUG:capa:     https://github.com/mandiant/capa-rules
DEBUG:capa:--------------------------------------------------------------------------------
DEBUG:capa:--------------------------------------------------------------------------------
DEBUG:capa: Using default embedded signatures.
DEBUG:capa: To provide your own signatures, use the form `capa.exe --signature ./path/to/signatures/  /path/to/mal.exe`.
DEBUG:capa:--------------------------------------------------------------------------------
DEBUG:capa:reading rules from directory c:\users\hp\documents\github\capa\capa\..\rules
DEBUG:capa.rules.cache:loading rule set from cache: C:\Users\HP\AppData\Local\flare\capa\cache\capa-10e51067.cache
DEBUG:capa:successfully loaded 797 rules
DEBUG:capa:analyzed file and extracted 131 features
DEBUG:capa:reading signatures from directory c:\users\hp\documents\github\capa\sigs
DEBUG:capa:found signature file: c:\users\hp\documents\github\capa\sigs\1_flare_msvc_rtf_32_64.sig
DEBUG:capa:found signature file: c:\users\hp\documents\github\capa\sigs\2_flare_msvc_atlmfc_32_64.sig
DEBUG:capa:found signature file: c:\users\hp\documents\github\capa\sigs\3_flare_common_libs.sig
DEBUG:viv_utils.idaloader:failed to import IDA Pro modules
DEBUG:capa:generating vivisect workspace for: .\tests\data\Practical Malware Analysis Lab 01-01.exe_
DEBUG:viv_utils.flirt:perf: flirt: parsing .sig: c:\users\hp\documents\github\capa\sigs\1_flare_msvc_rtf_32_64.sig: 0.48s
DEBUG:viv_utils.flirt:flirt: sig count: 210632
DEBUG:viv_utils.flirt:perf: flirt: compiling sigs: 1.27s
|DEBUG:viv_utils.flirt:registering viv function analyzer: FlirtFunctionAnalyzer (c:\users\hp\documents\github\capa\sigs\1_flare_msvc_rtf_32_64.sig)
DEBUG:viv_utils.flirt:perf: flirt: parsing .sig: c:\users\hp\documents\github\capa\sigs\2_flare_msvc_atlmfc_32_64.sig: 0.86s
DEBUG:viv_utils.flirt:flirt: sig count: 381783
DEBUG:viv_utils.flirt:perf: flirt: compiling sigs: 1.98s
/DEBUG:viv_utils.flirt:registering viv function analyzer: FlirtFunctionAnalyzer (c:\users\hp\documents\github\capa\sigs\2_flare_msvc_atlmfc_32_64.sig)
DEBUG:viv_utils.flirt:perf: flirt: parsing .sig: c:\users\hp\documents\github\capa\sigs\3_flare_common_libs.sig: 0.31s
DEBUG:viv_utils.flirt:flirt: sig count: 119337
DEBUG:viv_utils.flirt:perf: flirt: compiling sigs: 0.61s
-DEBUG:viv_utils.flirt:registering viv function analyzer: FlirtFunctionAnalyzer (c:\users\hp\documents\github\capa\sigs\3_flare_common_libs.sig)
 analyzing programDEBUG:viv_utils.flirt:found library function: 0x40193c: __setdefaultprecision
\ analyzing programDEBUG:capa:format: pe, platform: windows, architecture: i386, number of functions: 13
DEBUG:capa:CAPA_SAVE_WORKSPACE unset, not saving workspace
matching:   0%|                                                                                                                  | 0/13 [00:00<?, ? functions/s, skipped 0 library functions]DEBUG:capa:analyzed function 0x401000 and extracted 39 features
DEBUG:capa:analyzed function 0x401040 and extracted 25 features
DEBUG:capa:analyzed function 0x401070 and extracted 24 features
DEBUG:capa:analyzed function 0x4010a0 and extracted 113 features
matching:  31%|████████████████████████████████▌                                                                         | 4/13 [00:00<00:00, 35.55 functions/s, skipped 0 library functions]DEBUG:capa:analyzed function 0x4011e0 and extracted 104 features
DEBUG:capa:analyzed function 0x401440 and extracted 198 features
DEBUG:capa:analyzed function 0x401820 and extracted 65 features
DEBUG:capa:analyzed function 0x401930 and extracted 7 features
matching:  62%|█████████████████████████████████████████████████████████████████▏                                        | 8/13 [00:00<00:00, 22.92 functions/s, skipped 0 library functions]DEBUG:capa:analyzed function 0x401936 and extracted 7 features
DEBUG:capa:skipping library function 0x40193c (__setdefaultprecision)
matching:  62%|██████████████████████████████████████████████████████████████▏                                      | 8/13 [00:00<00:00, 22.92 functions/s, skipped 1 library functions (8%)]DEBUG:capa:analyzed function 0x40194e and extracted 6 features
DEBUG:capa:analyzed function 0x401951 and extracted 6 features
DEBUG:capa:analyzed function 0x401966 and extracted 7 features
matching: 100%|████████████████████████████████████████████████████████████████████████████████████████████████████| 13/13 [00:00<00:00, 35.42 functions/s, skipped 1 library functions (8%)] 
DEBUG:capa:analyzed file and extracted 135 features
+------------------------+------------------------------------------------------------------------------------+
| md5                    | bb7425b82141a1c0f7d60e5106676bb1                                                   |
| sha1                   | 9dce39ac1bd36d877fdb0025ee88fdaff0627cdb                                           |
| sha256                 | 58898bd42c5bd3bf9b1389f0eee5b39cd59180e8370eb9ea838a0b327bd6fe47                   |
| os                     | windows                                                                            |
| format                 | pe                                                                                 |
| arch                   | i386                                                                               |
| path                   | tests\data\Practical Malware Analysis Lab 01-01.exe_                               |
+------------------------+------------------------------------------------------------------------------------+

+------------------------+------------------------------------------------------------------------------------+
| ATT&CK Tactic          | ATT&CK Technique                                                                   |
|------------------------+------------------------------------------------------------------------------------|
| DISCOVERY              | File and Directory Discovery T1083                                                 |
+------------------------+------------------------------------------------------------------------------------+

+-----------------------------+-------------------------------------------------------------------------------+
| MBC Objective               | MBC Behavior                                                                  |
|-----------------------------+-------------------------------------------------------------------------------|
| DISCOVERY                   | Code Discovery::Enumerate PE Sections [B0046.001]                             |
|                             | File and Directory Discovery [E1083]                                          |
| FILE SYSTEM                 | Copy File [C0045]                                                             |
|                             | Read File [C0051]                                                             |
+-----------------------------+-------------------------------------------------------------------------------+

+------------------------------------------------------+------------------------------------------------------+
| CAPABILITY                                           | NAMESPACE                                            |
|------------------------------------------------------+------------------------------------------------------|
| copy file                                            | host-interaction/file-system/copy                    |
| enumerate files recursively                          | host-interaction/file-system/files/list              |
| read file via mapping (2 matches)                    | host-interaction/file-system/read                    |
| enumerate PE sections                                | load-code/pe                                         |
| resolve function by parsing PE exports               | load-code/pe                                         |
+------------------------------------------------------+------------------------------------------------------+
```

williballenthin commented:
> This should improve the output when other things are logged along the execution, e.g., when running in debug mode -d.

logging writes to stderr while this PR updates stdout. perhaps we should fix the logging output in another PR?

williballenthin commented:
to be clear, what im talking about is, today, we see:

```
DEBUG:capa:analyzed function 0x401040 and extracted 25 features
DEBUG:capa:analyzed function 0x401070 and extracted 24 features
DEBUG:capa:analyzed function 0x4010a0 and extracted 113 features
matching:  31%|████████████████████████████████▌                                                                         | 4/13 [00:00<00:00, 35.55 functions/s, skipped 0 library functions]DEBUG:capa:analyzed function 0x4011e0 and extracted 104 features
DEBUG:capa:analyzed function 0x401440 and extracted 198 features
DEBUG:capa:analyzed function 0x401820 and extracted 65 features
DEBUG:capa:analyzed function 0x401930 and extracted 7 features
matching:  62%|█████████████████████████████████████████████████████████████████▏                                        | 8/13 [00:00<00:00, 22.92 functions/s, skipped 0 library functions]DEBUG:capa:analyzed function 0x401936 and extracted 7 features
DEBUG:capa:skipping library function 0x40193c (__setdefaultprecision)
matching:  62%|██████████████████████████████████████████████████████████████▏                                      | 8/13 [00:00<00:00, 22.92 functions/s, skipped 1 library functions (8%)]DEBUG:capa:analyzed function 0x40194e and extracted 6 features
DEBUG:capa:analyzed function 0x401951 and extracted 6 features
DEBUG:capa:analyzed function 0x401966 and extracted 7 features
matching: 100%|████████████████████████████████████████████████████████████████████████████████████████████████████| 13/13 [00:00<00:00, 35.42 functions/s, skipped 1 library functions (8%)] 
DEBUG:capa:analyzed file and extracted 135 features
```

note that snapshots of the progress bar get left behind as logging messages are printed. we would want to see this:

```
DEBUG:capa:analyzed function 0x401040 and extracted 25 features
DEBUG:capa:analyzed function 0x401070 and extracted 24 features
DEBUG:capa:analyzed function 0x4010a0 and extracted 113 features
DEBUG:capa:analyzed function 0x4011e0 and extracted 104 features
DEBUG:capa:analyzed function 0x401440 and extracted 198 features
DEBUG:capa:analyzed function 0x401820 and extracted 65 features
DEBUG:capa:analyzed function 0x401930 and extracted 7 features
DEBUG:capa:analyzed function 0x401936 and extracted 7 features
DEBUG:capa:skipping library function 0x40193c (__setdefaultprecision)
DEBUG:capa:analyzed function 0x40194e and extracted 6 features
DEBUG:capa:analyzed function 0x401951 and extracted 6 features
DEBUG:capa:analyzed function 0x401966 and extracted 7 features
matching: 100%|████████████████████████████████████████████████████████████████████████████████████████████████████| 13/13 [00:00<00:00, 35.42 functions/s, skipped 1 library functions (8%)] 
```

but i think this will take a little more research to see if there logging + tqdm integration, or if we have to hook stderr like @Aayush-Goel-04 did in this PR.

williballenthin commented on scripts/lint.py:
```diff
@@ -907,7 +874,7 @@ def lint(ctx: Context):
     with tqdm.contrib.logging.tqdm_logging_redirect(source_rules, unit="rule") as pbar:
-        with redirecting_print_to_tqdm():
+        with capa.helpers.redirecting_print_to_tqdm(False):
             for rule in pbar:
```
this looks good, thank you!

Aayush-Goel-04 commented:
Sorry, I accidentally closed this PR, will repopen it

williballenthin commented:
> looks like we can probably use tqdm.contrib.logging.logging_redirect_tqdm. @Aayush-Goel-04 want to give this a shot?

if you have some time, would you try out these changes, too? happy to merge a new PR with this new functionality.
